doc/classes/MainLoop.xml

```diff
@@ -5,40 +5,53 @@
-		Upon the application start, a [MainLoop] implementation must be provided to the OS; otherwise, the application will exit. This happens automatically (and a [SceneTree] is created) unless a main [Script] is provided from the command line (with e.g. [code]godot -s my_loop.gd[/code], which should then be a [MainLoop] implementation.
+		Upon the application start, a [MainLoop] implementation must be provided to the OS; otherwise, the application will exit. This happens automatically (and a [SceneTree] is created) unless a [MainLoop] [Script] is provided from the command line (with e.g. [code]godot -s my_loop.gd[/code] or the "Main Loop Type" project setting is overwritten.
 		Here is an example script implementing a simple [MainLoop]:
-		[b]FIXME:[/b] No longer valid after DisplayServer split and Input refactoring.
 		[codeblock]
+		class_name CustomMainLoop
 		extends MainLoop
 
 		var time_elapsed = 0
-		var keys_typed = []
 		var quit = false
 
 		func _initialize():
 		    print("Initialized:")
 		    print("  Starting time: %s" % str(time_elapsed))
+		    _make_quit_action()
+
 
 		func _idle(delta):
 		    time_elapsed += delta
+
 		    # Return true to end the main loop.
-		    return quit
+		    return quit || Input.is_action_just_released("quit")
 
-		func _input_event(event):
-		    # Record keys.
-		    if event is InputEventKey and event.pressed and !event.echo:
-		        keys_typed.append(OS.get_keycode_string(event.keycode))
-		        # Quit on Escape press.
-		        if event.keycode == KEY_ESCAPE:
-		            quit = true
-		    # Quit on any mouse click.
-		    if event is InputEventMouseButton:
-		        quit = true
 
 		func _finalize():
 		    print("Finalized:")
 		    print("  End time: %s" % str(time_elapsed))
-		    print("  Keys typed: %s" % var2str(keys_typed))
+
+
+		# In your Main Loop you would have a viewport that you can use to gather input events
+		# You partly implement your quit logic in your viewport and then set quit to true to exit
+		func _make_quit_action():
+		    InputMap.add_action("quit")
+		    # Escape
+		    var esc = InputEventKey.new()
+		    esc.keycode = KEY_ESCAPE
+		    InputMap.action_add_event("quit", esc)
+		    # Left mouse button
+		    var mouse_button_left = InputEventMouseButton.new()
+		    mouse_button_left.button_index = BUTTON_LEFT
+		    InputMap.action_add_event("quit", mouse_button_left)
+		    # Right mouse button
+		    var mouse_button_right = InputEventMouseButton.new()
+		    mouse_button_right.button_index = BUTTON_RIGHT
+		    InputMap.action_add_event("quit", mouse_button_right)
+		    # Middle Mouse Button
+		    var mouse_button_middle = InputEventMouseButton.new()
+		    mouse_button_middle.button_index = BUTTON_MIDDLE
+		    InputMap.action_add_event("quit", mouse_button_middle)
```

Comment:
That's a bit overkill. In the example I used `if event is InputEventMouseButton` just to keep things simple, but this fully spelt out creation of actions makes things more complex than they need to be.